napari/_vispy/vendored/volume.py

```diff
@@ -201,6 +201,106 @@
     return final_color;
 }}
 
+vec4 calculateCategoricalColor(vec4 betterColor, vec3 loc, vec3 step)
+{{   
+    // Calculate color by incorporating lighting
+    vec4 color0 = $sample(u_volumetex, loc);
+    vec4 color1;
+    vec4 color2;
+    float val0 = colorToVal(color0);
+    float val1 = 0;
+    float val2 = 0;
+    int n_dark_axes = 0;
+    
+    // View direction
+    vec3 V = normalize(view_ray);
+    
+    // calculate normal vector from gradient
+    vec3 N; // normal
+    color1 = $sample( u_volumetex, loc+vec3(-step[0],0.0,0.0) );
+    color2 = $sample( u_volumetex, loc+vec3(step[0],0.0,0.0) );
+
+    N[0] = colorToVal(color1) - colorToVal(color2);
+    val1 = colorToVal(color1);
+    val2 = colorToVal(color2);
+    if (val1 < val0) {{
+        if (val2 < val0) {{
+            n_dark_axes = n_dark_axes + int(1);
+        }}
+        
+    }}
```

Comment:
This is the hack I mentioned in the PR description to detect voxels that protrude into the background, away from the surface). The reason we need to do this is that the surface normal estimation is bad for these pixels and thus the diffusive illumination isn't correct. This results in dim pixels on the edges of surfaces (see below). I am definitely open to other fixes.

![Screenshot 2021-07-09 at 15 45 06](https://user-images.githubusercontent.com/1120672/125087287-aae4be80-e0cc-11eb-811f-37ba2a382175.png)